### ISSUE TITLE
WP provides a TOC without leaving the resource

### ISSUE THREAD

**Issue by iherman** [w3c/wpub#146]
User agents should provide an affordance for accessing the table of contents without leaving the resource they are currently viewing.

**Comment by TzviyaSiegman** [comment]
I might tweak this to say that UA should provide access to TOC from anywhere in the publication.

**Comment by avneeshsingh** [comment]
+1
Access to TOC without leaving current resource is important for accessibility.

**Comment by rakutenjeff** [comment]
I think the policy for UA to provide affordance support is it can provide feature for reading mode, as well as not block Reading System build upon UA to customize affordance.

So, I believe UA should support TOC from following two ways but also can provide option to let Reading System disable displaying TOC.
1. provide TOC list as a global js object (for example navigator.toc)
2. provide UI support of TOC (when UA entered "Reading Mode", it should provide something like following image) 
![toc-menu](https://user-images.githubusercontent.com/4615101/39709385-ac799b48-51e7-11e8-81bb-c2c9027a4e3f.png)

How do you think about it?

**Comment by rakutenjeff** [comment]
Thanks @HadrienGardeur
Yes, agree, for above 1. having it inside `WebPublicationManifest` looks good to me.
Maybe define the TOC list as sequence of something. Also hope can define it as a tree structure (Or at least can give a sense about relationship between Chapter and section underneath).

Once we have consensus on this issue I will send PR to add it into `WebPublicationManifest`.

**Comment by HadrienGardeur** [comment]
We already use a sequence of `PublicationLink` dictionaries for `links`, `reading_order` and `resources`. A small tweak to `PublicationLink` could turn it into a tree structure:

```webidl
dictionary PublicationLink {
    required USVString                 href;
             DOMString                 type;
             USVString                 title;
             sequence<DOMString>       rel;
             sequence<PublicationLink> child;
             DOMString                 height;
             DOMString                 width;
             boolean                   templated;
};
```

This would work for pretty much any navigation that we can think of (TOC, landmarks, list of illustrations/videos/audio).

**Comment by Jeffxz** [comment]
Cool. thanks @HadrienGardeur. Yes will see how it works with TOC.

Anyone else has some feedback? How about 2. and 3. in https://github.com/w3c/wpub/issues/146#issuecomment-387099066 ?

**Comment by rakutenjeff** [comment]
Cool. thanks @HadrienGardeur. Yes will see how it works with TOC.

Anyone else has some feedback? How about 2. and 3. in #146 (comment) ?

**Comment by iherman** [comment]
Tzviya will make a writeup for this affordance, based on the discussion on 2018-05-31

**Comment by HadrienGardeur** [comment]
I'd like to clarify this a bit:

- we agreed in the F2F that the manifest will hint at the location of the ToC
- we've also agreed to express the ToC itself using HTML and identify it with `doc-toc`
- do we expect that the UA will process the HTML to have its own in-memory representation of the ToC?
- if so, should I add it to the WebIDL?

**Comment by iherman** [comment]
> I'd like to clarify this a bit:
> 
> we agreed in the F2F that the manifest will hint at the location of the ToC
yes
> we've also agreed to express the ToC itself using HTML and identify it with doc-toc
yes
> do we expect that the UA will process the HTML to have its own in-memory representation of the ToC?
we have not discussed this explicitly, although I would think so...

> if so, should I add it to the WebIDL?
At some point yes.

There were some discussion on what the role of IDL would be in general, but that has not been fully decided yet. I would think we should have a clear idea about that before we spend energy on it.

**Comment by HadrienGardeur** [comment]
> There were some discussion on what the role of IDL would be in general, but that has not been fully decided yet. I would think we should have a clear idea about that before we spend energy on it.

From an implementer's point of view, the IDL is extremely helpful since it provides a clear model how you should implement support for WP.

Our JSON-LD approach is going to be very flexible due to the nature of JSON, JSON-LD and schema.org.
UAs will need to process it and turn it into something canonical that it can easily work with.

**Comment by rakutenjeff** [comment]
I agree with Hadrien on we need toc to be added to IDL as data structure. We need discussion in weekly meeting though. The discussion about IDL might be different from this issue. I can split it to other issue if you guys also think it's better.

**Comment by llemeurfr** [comment]
IMO the (Web)IDL is the conceptual model of the Publication, in practice the memory object the UA will deal with. Whatever the serialization is, in order to fulfill need to display a ToC without leaving the current resource, the the UA will need to create an in-memory representation of the ToC as a sub-structure of the Publication object. And the ToC will need to be represented as a tree object. Plus, each ToC label will need to be represented as a String (the difference btw USVSString and DOMString in the WebIDL is still fuzzy for me).

**Comment by iherman** [comment]
O.k., reading comments, I am fine with having an IDL.

**Comment by TzviyaSiegman** [comment]
information for PR and for use cases:

The user agent should provide access to the table of contents without leaving current resource from anywhere in the publication.

The table of contents is a listed as a structural property in the infoset (https://w3c.github.io/wpub/#wp-table-of-contents). 

Dependency: TBD how the manifest identifies the table of contents.

For Accessibility reasons, many think that this should be a MUST for User Agents to allow multiple ways for users to access content.

Potential use cases to add to UCR: Yuuko is reading about reading about War World II. She is in middle of chapter 7 and opens the table of contents to scroll through it to understand how much more reading she has left.
Marie is taking a self-assement quiz in her Calculus book. She checks the table of contents to determine if the solutions are provided at a later point in the book.

If UA supports this it might provide a button from which to access the ToC that is accessible from all locations in the publication.

**Comment by iherman** [comment]
@TzviyaSiegman I have translated https://github.com/w3c/wpub/issues/146#issuecomment-397041646 into a PR (#226) with minor changes:

- I guess it is not possible to put a TOC as a MUST in this section, so I put RECOMMENDED
- I have added explicit references to the Use Case document requirement list
- I have not touched the additional example, I believe that should be added to the UCR document (@atyposh ?)

**Comment by TzviyaSiegman** [comment]
@iherman Thank you. I had planned to do the PR today. I agree with your changes and will work with @atyposh

**Comment by TzviyaSiegman** [comment]
@GeorgeKerscher adds that focus in TOC should be on the element from which the user access it. For example if the user accesses the TOC from chapter 10, focus in TOC should be on chapter 10 link.
